bugzoo/mgr/coverage/gcov.py

```diff
@@ -0,0 +1,283 @@
+        container = self.__container
+        mgr_ctr = self.__installation.containers
+        mgr_bug = self.__installation.bugs
+        mgr_file = self.__installation.files
+        bug = mgr_bug[container.bug]
+        dir_source = bug.source_dir
+
+        # add instrumentation to each file
+        for fn_src in self.__files_to_instrument:
+            fn_src = os.path.join(dir_source, fn_src)
+            logger.debug("instrumenting file [%s] in container [%s]",
+                         fn_src, container.uid)
+            contents_original = mgr_file.read(container, fn_src)
+            contents_instrumented = INSTRUMENTATION + contents_original
+            mgr_file.write(container, fn_src, contents_instrumented)
+
+        # recompile with instrumentation options
+        outcome = mgr_ctr.compile_with_instrumentation(container)
+        if not outcome.successful:
+            msg = "failed to generate coverage for container ({}) due to compilation failure."
+            msg = msg.format(container.id)
+            logger.debug("failed build output: %s", outcome.response.output)
+            raise Exception(msg)
+
+        logger.debug("instrumented container: %s", container.uid)
+
+    def run(self,
+            tests: Iterable[TestCase],
+            *,
+            instrument: bool = True
+            ) -> TestSuiteCoverage:
+        """
+        Uses a provided container to compute line coverage information for a
+        given list of tests.
+        """
+        logger.debug("computing coverage for container: %s", container.uid)
+        container = self.__container
+
+        try:
+            if instrument:
+                logger.debug("instrumenting container")
+                self.prepare()
+            else:
+                logger.debug("not instrumenting container")
+        except Exception:
+            msg = "failed to instrument container."
+            raise exceptions.FailedToComputeCoverage(msg)
+
+        cov = {}
+        for test in _tests:
+            logger.debug("Generating coverage for test %s in container %s",
+                         test.name, container.uid)
+            outcome = self.__installation.containers.execute(container, test)
+            filelines = self.extract()
+            test_coverage = TestCoverage(test.name, outcome, filelines)
+            logger.debug("Generated coverage for test %s in container %s",
+                         test.name, container.uid)
+            cov[test.name] = test_coverage
+
+        self.cleanup()
+
+        coverage = TestSuiteCoverage(cov)
+        logger.debug("Computed coverage for container: %s", container.uid)
+        return coverage
+
+    def cleanup(self) -> None:
+        """
+        Strips instrumentation from the source code inside the container,
+        and reconfigures its program to no longer use coverage options.
+        """
+        logger.debug("cleanup method for gcov extractor does nothing.")
```

Comment:
Shouldn't it deinstrument?

Reply:
It should deinstrument, but the code that we had for doing that didn't really work and was disabled accordingly. I'll open a separate issue for this.

Reply:
#302